// can-import-module/dist/amd/can-import-module.js
/*[process-shim]*/
(function(global, env) {
	// jshint ignore:line
	if (typeof process === "undefined") {
		global.process = {
			argv: [],
			cwd: function() {
				return "";
			},
			browser: true,
			env: {
				NODE_ENV: env || "development"
			},
			version: "",
			platform:
				global.navigator &&
				global.navigator.userAgent &&
				/Windows/.test(global.navigator.userAgent)
					? "win"
					: ""
		};
	}
})(
	typeof self == "object" && self.Object == Object
		? self
		: typeof process === "object" &&
		  Object.prototype.toString.call(process) === "[object process]"
			? global
			: window,
	"development"
);

/*[global-shim-start]*/
(function(exports, global, doEval) {
	// jshint ignore:line
	var origDefine = global.define;

	var get = function(name) {
		var parts = name.split("."),
			cur = global,
			i;
		for (i = 0; i < parts.length; i++) {
			if (!cur) {
				break;
			}
			cur = cur[parts[i]];
		}
		return cur;
	};
	var set = function(name, val) {
		var parts = name.split("."),
			cur = global,
			i,
			part,
			next;
		for (i = 0; i < parts.length - 1; i++) {
			part = parts[i];
			next = cur[part];
			if (!next) {
				next = cur[part] = {};
			}
			cur = next;
		}
		part = parts[parts.length - 1];
		cur[part] = val;
	};
	var useDefault = function(mod) {
		if (!mod || !mod.__esModule) return false;
		var esProps = { __esModule: true, default: true };
		for (var p in mod) {
			if (!esProps[p]) return false;
		}
		return true;
	};

	var hasCjsDependencies = function(deps) {
		return (
			deps[0] === "require" && deps[1] === "exports" && deps[2] === "module"
		);
	};

	var modules =
		(global.define && global.define.modules) ||
		(global._define && global._define.modules) ||
		{};
	var ourDefine = (global.define = function(moduleName, deps, callback) {
		var module;
		if (typeof deps === "function") {
			callback = deps;
			deps = [];
		}
		var args = [],
			i;
		for (i = 0; i < deps.length; i++) {
			args.push(
				exports[deps[i]]
					? get(exports[deps[i]])
					: modules[deps[i]] || get(deps[i])
			);
		}
		// CJS has no dependencies but 3 callback arguments
		if (hasCjsDependencies(deps) || (!deps.length && callback.length)) {
			module = { exports: {} };
			args[0] = function(name) {
				return exports[name] ? get(exports[name]) : modules[name];
			};
			args[1] = module.exports;
			args[2] = module;
		}
		// Babel uses the exports and module object.
		else if (!args[0] && deps[0] === "exports") {
			module = { exports: {} };
			args[0] = module.exports;
			if (deps[1] === "module") {
				args[1] = module;
			}
		} else if (!args[0] && deps[0] === "module") {
			args[0] = { id: moduleName };
		}

		global.define = origDefine;
		var result = callback ? callback.apply(null, args) : undefined;
		global.define = ourDefine;

		// Favor CJS module.exports over the return value
		result = module && module.exports ? module.exports : result;
		modules[moduleName] = result;

		// Set global exports
		var globalExport = exports[moduleName];
		if (globalExport && !get(globalExport)) {
			if (useDefault(result)) {
				result = result["default"];
			}
			set(globalExport, result);
		}
	});
	global.define.orig = origDefine;
	global.define.modules = modules;
	global.define.amd = true;
	ourDefine("@loader", [], function() {
		// shim for @@global-helpers
		var noop = function() {};
		return {
			get: function() {
				return { prepareGlobal: noop, retrieveGlobal: noop };
			},
			global: global,
			__exec: function(__load) {
				doEval(__load.source, global);
			}
		};
	});
})(
	{},
	typeof self == "object" && self.Object == Object
		? self
		: typeof process === "object" &&
		  Object.prototype.toString.call(process) === "[object process]"
			? global
			: window,
	function(__$source__, __$global__) {
		// jshint ignore:line
		eval("(function() { " + __$source__ + " \n }).call(__$global__);");
	}
);

/*can-import-module@1.3.1#loader/steal-optimized*/
define('can-import-module@1.3.1#loader/steal-optimized', [
    'require',
    'exports',
    'module',
    'can-globals@1.2.2#global/global'
], function (require, exports, module) {
    (function (global, require, exports, module) {
        var getGlobal = require('can-globals@1.2.2#global/global');
        var global = getGlobal();
        module.exports = function (moduleName, parentName) {
            if (typeof global.stealRequire !== 'undefined') {
                return steal.import(moduleName, { name: parentName });
            }
        };
    }(function () {
        return this;
    }(), require, exports, module));
});
/*can-import-module@1.3.1#loader/system*/
define('can-import-module@1.3.1#loader/system', [
    'require',
    'exports',
    'module',
    'can-globals@1.2.2#global/global'
], function (require, exports, module) {
    (function (global, require, exports, module) {
        var getGlobal = require('can-globals@1.2.2#global/global');
        var global = getGlobal();
        function isFunction(fn) {
            return typeof fn === 'function';
        }
        module.exports = function (moduleName, parentName) {
            if (typeof global.System === 'object' && isFunction(global.System['import'])) {
                return global.System['import'](moduleName, { name: parentName });
            }
        };
    }(function () {
        return this;
    }(), require, exports, module));
});
/*can-import-module@1.3.1#loader/es6*/
define('can-import-module@1.3.1#loader/es6', [
    'require',
    'exports',
    'module',
    'can-globals@1.2.2#global/global'
], function (require, exports, module) {
    (function (global, require, exports, module) {
        var getGlobal = require('can-globals@1.2.2#global/global');
        if ('HTMLScriptElement' in getGlobal() && 'noModule' in HTMLScriptElement.prototype) {
            module.exports = new Function('moduleName', 'if (!(moduleName.match(/[^\\\\\\/]\\.([^.\\\\\\/]+)$/) || [null]).pop()) {\n' + 'moduleName += \'.js\';\n' + '}\n' + 'return import(moduleName.replace(/[\'"]+/g, \'\'));\n');
        } else {
            module.exports = function () {
            };
        }
    }(function () {
        return this;
    }(), require, exports, module));
});
/*can-import-module@1.3.1#loader/node*/
define('can-import-module@1.3.1#loader/node', [
    'require',
    'exports',
    'module',
    'can-globals@1.2.2#is-node/is-node'
], function (require, exports, module) {
    (function (global, require, exports, module) {
        var isNode = require('can-globals@1.2.2#is-node/is-node');
        module.exports = function (moduleName) {
            if (isNode()) {
                return Promise.resolve(require(moduleName));
            }
        };
    }(function () {
        return this;
    }(), require, exports, module));
});
/*can-import-module@1.3.1#loader/require*/
define('can-import-module@1.3.1#loader/require', [
    'require',
    'exports',
    'module',
    'can-globals@1.2.2#global/global'
], function (require, exports, module) {
    (function (global, require, exports, module) {
        var getGlobal = require('can-globals@1.2.2#global/global');
        var global = getGlobal();
        module.exports = function (moduleName) {
            if (global.define && global.define.amd) {
                return new Promise(function (resolve, reject) {
                    global.require([moduleName], function (value) {
                        resolve(value);
                    });
                });
            }
        };
    }(function () {
        return this;
    }(), require, exports, module));
});
/*can-import-module@1.3.1#can-import-module*/
define('can-import-module@1.3.1#can-import-module', [
    'require',
    'exports',
    'module',
    'can-namespace@1.0.0#can-namespace',
    'can-import-module@1.3.1#loader/steal-optimized',
    'can-import-module@1.3.1#loader/system',
    'can-import-module@1.3.1#loader/es6',
    'can-import-module@1.3.1#loader/node',
    'can-import-module@1.3.1#loader/steal-optimized',
    'can-import-module@1.3.1#loader/es6',
    'can-import-module@1.3.1#loader/node',
    'can-import-module@1.3.1#loader/require',
    'can-import-module@1.3.1#loader/system'
], function (require, exports, module) {
    'use strict';
    var namespace = require('can-namespace@1.0.0#can-namespace');
    var loader = [];
    function addLoader(fn) {
        if (typeof fn === 'function') {
            loader.push(fn);
        }
    }
    function flushLoader() {
        loader = [];
    }
    function preset(preset) {
        flushLoader();
        switch (preset) {
        case 'stealjs':
            addLoader(require('can-import-module@1.3.1#loader/steal-optimized'));
            addLoader(require('can-import-module@1.3.1#loader/system'));
            break;
        case 'ES2020':
        case 'es2020':
        case 'dynamic-import':
            addLoader(require('can-import-module@1.3.1#loader/es6'));
            break;
        case 'node':
            addLoader(require('can-import-module@1.3.1#loader/node'));
            break;
        case 'all':
        default:
            addLoader(require('can-import-module@1.3.1#loader/steal-optimized'));
            addLoader(require('can-import-module@1.3.1#loader/es6'));
            addLoader(require('can-import-module@1.3.1#loader/node'));
            addLoader(require('can-import-module@1.3.1#loader/require'));
            addLoader(require('can-import-module@1.3.1#loader/system'));
            break;
        }
    }
    preset('all');
    module.exports = namespace.import = function (moduleName, parentName) {
        return new Promise(function (resolve, reject) {
            try {
                var loaderPromise;
                for (var i = loader.length - 1; i >= 0; i--) {
                    loaderPromise = loader[i](moduleName, parentName);
                    if (loaderPromise) {
                        break;
                    }
                }
                if (loaderPromise) {
                    loaderPromise.then(resolve, reject);
                } else {
                    reject('no proper module-loader available');
                }
            } catch (err) {
                reject(err);
            }
        });
    };
    module.exports.addLoader = addLoader;
    module.exports.flushLoader = flushLoader;
    module.exports.preset = preset;
});
/*[global-shim-end]*/
(function(global) { // jshint ignore:line
	global._define = global.define;
	global.define = global.define.orig;
}
)(typeof self == "object" && self.Object == Object ? self : (typeof process === "object" && Object.prototype.toString.call(process) === "[object process]") ? global : window);
